plutarch-extra/Plutarch/Extra/Ord.hs

```diff
@@ -0,0 +1,681 @@
+-- Needed to 'link' Ordering and POrdering
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Plutarch.Extra.Ord (
+  -- * Types
+  POrdering (..),
+  PComparator,
+
+  -- * Functions
+
+  -- ** Creating comparators
+  pfromOrd,
+  pfromOrdBy,
+
+  -- ** Transforming comparators
+  pmapComparator,
+  preverseComparator,
+
+  -- ** Using comparators
+
+  -- *** Basic
+  pcompareBy,
+  pequateBy,
+  pleqBy,
+  pgeqBy,
+
+  -- *** Sortedness checking
+  pisSortedBy,
+
+  -- *** Uniqueness checking
+  pallUnique,
+  pallUniqueBy,
+  ptryAllUnique,
+  ptryAllUniqueBy,
+
+  -- *** Sorted merging
+  ptryMerge,
+  ptryMergeBy,
+
+  -- *** Sorting
+  psort,
+  psortBy,
+
+  -- *** Nubbing
+  pnubSort,
+  pnubSortBy,
+) where
+
+import Data.Semigroup (Semigroup (stimes), stimesIdempotentMonoid)
+import Plutarch.Extra.List (precListLookahead)
+import Plutarch.Extra.Maybe (ptraceIfNothing)
+import Plutarch.Internal.PlutusType (PlutusType (pcon', pmatch'))
+import Plutarch.Lift (
+  PConstantDecl (
+    PConstantRepr,
+    PConstanted,
+    pconstantFromRepr,
+    pconstantToRepr
+  ),
+  PUnsafeLiftDecl (PLifted),
+ )
+import Plutarch.List (pconvertLists)
+import Plutarch.Prelude
+import Plutarch.Unsafe (punsafeCoerce)
+
+{- | A representation of a comparison at the Plutarch level. Equivalent to
+ 'Ordering' in Haskell.
+-}
+data POrdering (s :: S)
+  = PLT
+  | PEQ
+  | PGT
+  deriving stock (Show, Generic)
+  deriving anyclass (PShow, PPartialOrd, POrd, PEq)
+
+instance PUnsafeLiftDecl POrdering where
+  type PLifted POrdering = Ordering
+
+instance PConstantDecl Ordering where
+  type PConstantRepr Ordering = Integer
+  type PConstanted Ordering = POrdering
+  pconstantToRepr = \case
+    LT -> 0
+    EQ -> 1
+    GT -> 2
+  pconstantFromRepr = \case
+    0 -> pure LT
+    1 -> pure EQ
+    2 -> pure GT
+    _ -> Nothing
+
+instance PlutusType POrdering where
+  type PInner POrdering = PInteger
```

Comment:
why

Reply:
Should this use SocttPlut?

Reply:
Possibly. I did the first thing that could possibly work, which seemed to perform OK, and actually allowed some optimizations (which I have measurements for).

Reply:
This is less efficient to match on and not really cheaper to construct. Scott-encoding is probably better.